RFE and RFECV allow features_to_select to be larger than available

DuarteSJ opened the pull request:
fixes #28725

Implemented a warning message for when the RFE or the RFECV objects are initialized with a n_features_to_select or min_features_to_select (respectively) attribute larger than the number of features present in the X variable that is passed to the fit method.

Additionally, a new test function has been added to ensure the correct warning message is raised in these scenarios. The test comprehensively covers both cases and validates that the warning is emitted as expected.

DuarteSJ committed:
RFE and RFECV allow features_to_select to be larger than available features (scikit-learn#28725)

Implemented a warning message for when the RFE or the RFECV objects are initialized with a n_features_to_select or min_features_to_select (respectively) attribute larger than the number of features present in the X variable that is passed to the fit method.

Additionally, a new test function has been added to ensure the correct warning message is raised in these scenarios. The test comprehensively covers both cases and validates that the warning is emitted as expected.

lesteve commented:
The bot comment https://github.com/scikit-learn/scikit-learn/pull/28764#issuecomment-2035135013 is trying to help. Apparently it was not that helpful for you, so if you see some way to improve it, let us know!

In the mean time, the easiest solution is probably to follow the instructions from the message above:
- install `ruff` 0.3.5
- run `ruff --fix --show-source .`
- commit the changes and push

Alternatively install `pre-commit` and run `precommit install`. When you do this on each commit, linting tools are run automatically (mostly `black` and `ruff`). If you are curious, you can see more details at https://pre-commit.com/.

DuarteSJ committed:
Fixed linting issues

ArturoAmorQ commented on sklearn/feature_selection/_rfe.py:
```diff
@@ -289,6 +290,15 @@ def _fit(self, X, y, step_score=None, **fit_params):
+        if n_features_to_select > n_features:
+            warnings.warn(
+                (
+                    "features_to_select shouldn't be greater than the number of"
+                    " available features"
```
We need to mention the actual behavior, what about something similar to

`n_features_to_select must not exceed available features. Falling back to select all n_features.`

ArturoAmorQ reviewed (commented):
Thanks for the PR @DuarteSJ, don't forget to also address the issue for `min_features_to_select` in `RFECV`.

DuarteSJ committed:
Improved the warning message to accurately describe the behavior. Moved the raising of the warning inside the scenario where n_features_to_select is already an integer.

DuarteSJ commented on sklearn/feature_selection/_rfe.py:
```diff
@@ -289,6 +290,15 @@ def _fit(self, X, y, step_score=None, **fit_params):
         else:  # float
             n_features_to_select = int(n_features * self.n_features_to_select)
 
+        if n_features_to_select > n_features:
```
I refrained from doing this because the fit method of this class already invokes the fit method of the RFE class. Do you want me to do it anyways?

ArturoAmorQ replied:
> Do you want me to do it anyways?

Yes, please. You can use a context manager to silence the warning when calling the RFE object inside the RFECV. Else the user would get the message repeated as many times as the number of cv splits, which can become very verbose really fast.

The other advantage is related to @jeremiedb's  https://github.com/scikit-learn/scikit-learn/pull/28764#discussion_r1553267276, as I would rather have the warning message to refer the correct hyperparameter `min_features_to_select` in `RFECV`.

jeremiedbb replied:
> Else the user would get the message repeated as many times as the number of cv splits, which can become very verbose really fast.

Indeed, I missed that. Then we should definitely raise the warning directly in RFECV and pass the curated value to RFE to not trigger the RFE warning.

DuarteSJ committed:
fixed test coverage

jeremiedbb commented on sklearn/feature_selection/_rfe.py:
```diff
@@ -286,6 +287,14 @@ def _fit(self, X, y, step_score=None, **fit_params):
+            if n_features_to_select > n_features:
+                warnings.warn(
+                    (
+                        "n_features_to_select must not exceed available features."
+                        " Falling back to select all n_features."
```
the "must not" would be appropriate for an error but feels weird when it's just a warning. What do you think of this rephrasing ?
```suggestion
                        f"Found {n_features_to_select=} > {n_features=}."
                        " There will be no feature selection and all features will be kept."
```

jeremiedbb commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -649,3 +650,44 @@ def test_rfe_estimator_attribute_error():
+    Non-regression test for:
+    https://github.com/scikit-learn/scikit-learn/issues/28725
+    """
+    X, y = make_classification(
+        n_samples=1000, n_features=20, n_redundant=0, n_classes=2, random_state=0
+    )
```
Let's not add variables irrelevant w.r.t the thing we're testing.
```suggestion
    X, y = make_classification(n_features=20, random_state=0)
```

jeremiedbb commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -649,3 +650,44 @@ def test_rfe_estimator_attribute_error():
+        rfe = ClsRFE(
+            estimator=LogisticRegression(random_state=0),
+            **{parameter_name: 21},
+            step=2,
```
same
```suggestion
```

jeremiedbb commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -649,3 +650,44 @@ def test_rfe_estimator_attribute_error():
+
+    # Catch the warning
+    with warnings.catch_warnings(record=True) as w:
+        # Determine the correct parameter name based on the estimator
+        if issubclass(ClsRFE, RFECV):
+            parameter_name = "min_features_to_select"
+        elif issubclass(ClsRFE, RFE):
+            parameter_name = "n_features_to_select"
+
```
Instead you could just modify the test parametrization with
```py
@pytest.mark.parametrize(
    "ClsRFE, param", [(RFE, "n_features_to_select"), (RFECV, "min_features_to_select")]
)
```

jeremiedbb commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -649,3 +650,44 @@ def test_rfe_estimator_attribute_error():
+        n_samples=1000, n_features=20, n_redundant=0, n_classes=2, random_state=0
+    )
+
+    # Catch the warning
+    with warnings.catch_warnings(record=True) as w:
```
`pytest` provides a more convenient way to assert that the warning is raised:
```suggestion
    with pytest.warns(UserWarning, match="n_features_to_select=21 > n_features=20"):
```

jeremiedbb commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -649,3 +650,44 @@ def test_rfe_estimator_attribute_error():
+        )
+        rfe.fit(X=X, y=y)
+
+        # Check if any UserWarning was raised
+        assert any(issubclass(warning.category, UserWarning) for warning in w)
+        # Check if the warning message is correct
+        assert any(
+            "n_features_to_select must not exceed available features."
+            " Falling back to select all n_features."
+            in str(warning.message)
+            for warning in w
+        )
```
None of that is needed if you use the pytest.warns context manager
```suggestion
```

jeremiedbb commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -649,3 +650,44 @@ def test_rfe_estimator_attribute_error():
+        assert any(issubclass(warning.category, UserWarning) for warning in w)
+        # Check if the warning message is correct
+        assert any(
+            "n_features_to_select must not exceed available features."
```
It's a bit weird that the warning will refer to `n_features_to_select` when using `RFECV`. I wonder if we should directly raise the warning from RFECV for it to be about `min_features_to_select`, but maybe that's good enough.

jeremiedbb reviewed (commented):
Thanks for the PR @DuarteSJ. Here are some comments

DuarteSJ committed:
Refactored warning messages and test cases for RFE and RFECV

- Updated the warning message in Recursive Feature Elimination (RFE) to specifically address the parameter `n_features_to_select`.
- Implemented a similar warning in Recursive Feature Elimination with Cross-Validation (RFECV) for the parameter `min_features_to_select`, excluding the warning when RFE is called from this function.
- Refactored the test cases to align with these changes, introducing parameterization to eliminate unnecessary conditional statements and removing unused variables.

DuarteSJ commented:
Thank you for all the help! Now I am wondering if it wouldn't be preferable to create two separate functions for testing RFE and RFECV, considering the need for numerous parameters in the test when they are joined. Separating the tests could improve code organization and readability. What are your thoughts on this?

jeremiedbb commented on sklearn/feature_selection/_rfe.py:
```diff
@@ -729,6 +738,15 @@ def fit(self, X, y, groups=None):
+            )
         rfe = RFE(
             estimator=self.estimator,
             n_features_to_select=self.min_features_to_select,
```
We can pass a curated `n_features_to_select` so that we're sure the warning won't be raised again in the inner RFE.
```suggestion
            n_features_to_select=min(self.min_features_to_select, n_features),
```

jeremiedbb commented on sklearn/feature_selection/_rfe.py:
```diff
@@ -729,6 +738,15 @@ def fit(self, X, y, groups=None):
+        if self.min_features_to_select > n_features:
+            warnings.warn(
+                (
+                    f"Found {self.min_features_to_select=} > {n_features=}. There will"
+                    " be no feature selection and all features will be kept."
```
Let's not include the `self`
```suggestion
                    f"Found min_features_to_select={self.min_features_to_select} > "
                    f"{n_features=}. There will be no feature selection and all "
                    "features will be kept."
```

jeremiedbb commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -649,3 +649,47 @@ def test_rfe_estimator_attribute_error():
+        )
+        rfe.fit(X=X, y=y)
+
+        # Check if the warning message is correct
+        assert any(warning_msg in str(warning.message) for warning in record)
```
That's not necessary. this is done by the pytest.warns context manager.
```suggestion
```

jeremiedbb commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -649,3 +649,47 @@ def test_rfe_estimator_attribute_error():
+    """
+    X, y = make_classification(n_features=20, random_state=0)
+
+    with pytest.warns(UserWarning, match=f"{param}=21 > n_features=20") as record:
```
Checking that part of the message is enough, no need to add the full message in the parametrization.
```suggestion
    with pytest.warns(UserWarning, match=f"{param}=21 > n_features=20"):
```

DuarteSJ committed:
applied sugestions

jeremiedbb commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -649,3 +649,26 @@ def test_rfe_estimator_attribute_error():
+    """Check if the correct warning is raised when trying to initialize a RFE
+    object with a n_features_to_select attribute larger than the number of
+    features present in the X variable that is passed to the fit method
+
+    Non-regression test for:
+    https://github.com/scikit-learn/scikit-learn/issues/28725
```
This PR does not fix a regression
```suggestion
```

jeremiedbb reviewed (commented):
Here's a last round of comments. Looks good otherwise.

DuarteSJ committed:
Refactor _rfe to remove unnecessary warning suppression

ArturoAmorQ commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -649,3 +649,23 @@ def test_rfe_estimator_attribute_error():
+    with pytest.warns(UserWarning, match=f"{param}=21 > n_features=20"):
+        # Create RFE/RFECV with n_features_to_select/min_features_to_select
+        # larger than the number of features present in the X variable
+        rfe = ClsRFE(
+            estimator=LogisticRegression(random_state=0),
+            **{param: 21},
+        )
+        rfe.fit(X=X, y=y)
```
I don't think setting the `random_state` in the logistic regression is relevant here as we are using the lbfgs solver by default.

```suggestion
        clsrfe = ClsRFE(
            estimator=LogisticRegression(),
            **{param: 21},
        )
        clsrfe.fit(X, y)
```

ArturoAmorQ reviewed (approved):
Just a couple of comments, but otherwise LGTM :)

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/DuarteSJ/28764

jeremiedbb committed:
address remaining review comments

jeremiedbb reviewed (approved):
I applied the last review comments. LGTM. Thanks @DuarteSJ

jeremiedbb committed:
Merge branch 'main' into DuarteSJ_RFE_and_RFECV_allow_features_to_select_to_be_larger_than_available_features
